### PR TITLE
attestations: Add plumbing for attestations

### PR DESCRIPTION
**Pull request by adityasaky**
This PR adds the basic plumbing for attestations from #164.

### CODE REVIEW

**Comment by adityasaky** [comment]
I'll add more tests for the authorizations specific APIs.

File: internal/attestations/attestations.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		return err
+	}
+
+	_, err = gitinterface.Commit(repo, treeHash, Ref, initialCommitMessage, false)
+	return err
+}
+
+// Attestations tracks all the attestations in a gittuf repository.
+type Attestations struct {
+	authorizations map[string]plumbing.Hash // refPath/from-to -> blob ID
```
[the remark this answers is not in the document]
Comment (wlynch) [reply internal/attestations/attestations.go:48]: It would be also useful to add more documentation (either README.md or godoc) on the Attestation format and how refs, commits, ReferenceEntries, and attestations relate to each other.

File: internal/attestations/authorization.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	predicateStruct, err := structpb.NewStruct(*predicateInterface)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ita.Statement{
+		Type: ita.StatementTypeUri,
+		Subject: []*ita.ResourceDescriptor{
+			{
+				Digest: map[string]string{digestGitCommitKey: toTargetID},
```
Comment (wlynch) [inline internal/attestations/authorization.go:67]:
```suggestion
				Digest: map[string]string{"sha1": toTargetID},
```

This should map to the algorithm, no? https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md

File: internal/attestations/authorization.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+type Authorization struct {
+	TargetRef    string `json:"targetRef"`
+	FromTargetID string `json:"fromTargetID"`
+	ToTargetID   string `json:"toTargetID"`
+}
+
+// NewAuthorizationAttestation creates a new authorization for the provided
+// information. The authorization is embedded in an in-toto "statement" and
+// returned with the appropriate "predicate type" set.
+func NewAuthorizationAttestation(targetRef, fromTargetID, toTargetID string) (*ita.Statement, error) {
```
Comment (wlynch) [inline internal/attestations/authorization.go:41]: Document what we mean by target id - I assume this means commit sha?

File: internal/attestations/authorization.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+		return ErrAuthorizationNotFound
+	}
+
+	delete(a.authorizations, authPath)
+	return nil
+}
+
+// GetAuthorizationAttestationFor returns the requested authorization
+// attestation (with its signatures).
+func (a *Attestations) GetAuthorizationAttestationFor(repo *git.Repository, refName, fromID, toID string) (*sslibdsse.Envelope, error) {
```
Comment (wlynch) [inline internal/attestations/authorization.go:115]: We probably need to tweak how we're storing attestations - I'd expect there to possibly be multiple attestations present for a given commit to/from pair. We can't rely on a single attestation per pair since these may need to be signed by different identities.

File: internal/attestations/authorization.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return err
+	}
+
+	attestation := &ita.Statement{}
+	if err := json.Unmarshal(payload, attestation); err != nil {
+		return err
+	}
+
+	if attestation.Subject[0].Digest["gitCommit"] != toID {
```
Comment (wlynch) [inline internal/attestations/authorization.go:155]: I'd iterate through here, since there could be multiple digests.

File: internal/attestations/authorization.go (hunk 4 above)
Comment (wlynch) [reply internal/attestations/authorization.go:115]: I also don't know if we should be keying on ref, since I would expect "give me all attestations for this commit" to be a fairly common workflow - having to scan all refs for possible commits would be costly as the repo grows.

Comment (wlynch) [inline internal/attestations/authorization.go:115]: We should add documentation about merge branches and the expected behavior of attestations.

```mermaid
    gitGraph
       commit id: "A"
       branch develop
       commit id:"B"
       commit id:"C"
       checkout main
       merge develop id:"D"
       commit id:"E"
```

e.g. If there's an attestation added to `develop:B->C`, this means that when it's merged to `main` the attestation won't be searchable via `main`, even though the commits are now part of the `main` branch. It makes sense when you realize that Authorization attestations were created w.r.t. a particular ref, but that may not be something that's immediately obvious.

**Review by wlynch — CHANGES_REQUESTED** [review]

File: internal/attestations/authorization.go (hunk 5 above)
Comment (naveensrinivasan) [inline internal/attestations/authorization.go:155]: This could potentially cause a panic in Go. If attestation.Subject is an empty slice.

File: internal/attestations/authorization.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	if predicate["fromTargetID"] != fromID {
+		return ErrInvalidAuthorization
+	}
+
+	if predicate["targetRef"] != refName {
+		return ErrInvalidAuthorization
+	}
+
```
Comment (naveensrinivasan) [inline internal/attestations/authorization.go:172]: The issue here is that the code assumes that the values in the predicate map are strings.

 If they are not, this will cause a panic.

File: internal/attestations/attestations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/attestations/attestations.go:48]: Updated the comment to be more complete of the specific implementation. I'll add authorizations more generally to the design doc, overriding #68.

File: internal/attestations/attestations.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	attestationsRootTree, err := gitinterface.GetTree(repo, attestationsCommit.TreeHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(attestationsRootTree.Entries) == 0 {
+		// This happens in the initial commit for the attestations namespace,
+		// with no entries in the tree at all
+		// An RSL entry isn't created by default for that entry but if one is,
+		// we don't want to panic.
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/attestations/attestations.go:87]: Do you mean the `Commit()` and other methods on `Attestations`?

File: internal/attestations/attestations.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+	attestations := &Attestations{authorizations: map[string]plumbing.Hash{}}
+
+	filesIter := authorizationsTree.Files()
+	if err := filesIter.ForEach(func(f *object.File) error {
+		attestations.authorizations[f.Name] = f.Blob.Hash
+		return nil
+	}); err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/attestations/attestations.go:111]: You're right, this is expensive. I hadn't peeked enough but this reads the blobs in questions. I was expressly trying to avoid this and just get the filenames by reading the trees rather than the blobs as well, reading the blob in question when the specific authorization is needed. Let me see how I can rework this.

File: internal/attestations/authorization.go (hunk 2 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:67]: https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#gitcommit-gittree-gitblob-gittag

The spec supports `git` specific objects as well for an additional hint!

File: internal/attestations/authorization.go (hunk 3 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:41]: Yes, most of the time. This is in-line with a recent RSL change where we switched from calling it `commitID` to `targetID` to support pointing to tag objects as well. I'll update the documentation.

File: internal/attestations/authorization.go (hunk 4 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:115]: > We probably need to tweak how we're storing attestations - I'd expect there to possibly be multiple attestations present for a given commit to/from pair. We can't rely on a single attestation per pair since these may need to be signed by different identities.

On this one, I disagree. In the `repository` package, the handler for adding an authorization for the same action when someone else signed one loads the attestation that exists and attaches a second signature to the DSSE envelope. That workflow isn't in this PR though. The predicate doesn't change, so that ought to be okay?

> I also don't know if we should be keying on ref, since I would expect "give me all attestations for this commit" to be a fairly common workflow - having to scan all refs for possible commits would be costly as the repo grows.

That's a good point and needs more thinking about. As a start I modeled this on the RSL reference entry as these are essentially additional signatures for an RSL. In the RSL we don't need to explicitly track the from ID, which is the only difference in the recorded information. I'll think about the data model some more.

File: internal/attestations/authorization.go (hunk 5 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:155]: Happy to add that here. However, do we foresee allowing users to attach attestations generated outside of gittuf for these workflows? The reason I didn't add this is because I expected gittuf to be the sole producer.

Comment (adityasaky) [reply internal/attestations/authorization.go:155]: Related: https://github.com/gittuf/gittuf/pull/170#discussion_r1390289843

Right now I expected gittuf to be the sole producer of this attestation, so we could be assured of the fields. But more checks may be in order depending on what we decide in the thread. Thanks for flagging this!

File: internal/attestations/authorization.go (hunk 6 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:172]: Also related to whether we can assume we're the creators of this specific attestation (which also has a gittuf.dev predicate type).

File: internal/attestations/authorization.go (hunk 4 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:120]: I _think_ that's fine? The authorization is for `develop` to move from B to C. On the `main` branch, the authorization (if one's needed) is to move from `A` to `D`. For a parallel, if we map "authorization" to PR approval, we care that the PR moving `main` from A to D was approved and less so a sub-PR to the feature branch?

Comment (adityasaky) [reply internal/attestations/authorization.go:120]: Note that this is _strictly_ for authorizations as they work now. Perhaps it needs to be called `reference-state-authorization` or something to make it more clear?

File: internal/attestations/attestations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/attestations/attestations.go]: https://github.com/gittuf/gittuf/pull/195

File: internal/attestations/authorization.go (hunk 4 above)
Comment (adityasaky) [reply internal/attestations/authorization.go:120]: https://github.com/gittuf/gittuf/pull/195

I've documented this better here, it might help clarify why we may not want this to move with changes being merged / subsumed into other branches.

File: internal/attestations/attestations.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/gittuf/gittuf/internal/gitinterface"
+	"github.com/gittuf/gittuf/internal/rsl"
+	"github.com/gittuf/gittuf/internal/third_party/go-git"
+	"github.com/gittuf/gittuf/internal/third_party/go-git/plumbing"
+	"github.com/gittuf/gittuf/internal/third_party/go-git/plumbing/filemode"
+	"github.com/gittuf/gittuf/internal/third_party/go-git/plumbing/object"
+)
+
+const (
+	Ref                                  = "refs/gittuf/attestations"
```
Comment (wlynch) [inline internal/attestations/attestations.go:17]: Non blocking, but I think we should think about how we can get this to play nice with existing git attestations like [`gitsign attest`](https://github.com/sigstore/gitsign/blob/f915adc8d4b3ccc417c322071d3dbb4bbd3bf1da/internal/commands/attest/README.md) 🤔  (I'll take an AI to dig into this more)

Comment (wlynch) [reply internal/attestations/attestations.go:17]: In particular - we should figure out sooner rather than later on breaking name changes (e.g. `refs/gittuf/attestations` vs `refs/attestations/commits`) as well as breaking storage formats.

File: internal/attestations/attestations.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (wlynch) [reply internal/attestations/attestations.go:94]: I think my original thought was that this was an unexpected case, but re-reading it seems like this might just be the initial state?

maybe rephrase this as:

```go
	if len(attestationsRootTree.Entries) == 0 {
		// This happens in the initial commit for the attestations namespace,
		// where there are no entries in the tree yet.
		// This is expected, and there is nothing more to check so return a zero Attestations state.
		return &Attestations{}, nil
	}
```

(added as suggestion)

File: internal/attestations/attestations.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if len(attestationsRootTree.Entries) == 0 {
+		// This happens in the initial commit for the attestations namespace,
+		// with no entries in the tree at all
+		// An RSL entry isn't created by default for that entry but if one is,
+		// we don't want to panic.
+		return &Attestations{}, nil
```
Comment (wlynch) [inline internal/attestations/attestations.go:95]:
```suggestion
	if len(attestationsRootTree.Entries) == 0 {
		// This happens in the initial commit for the attestations namespace,
		// where there are no entries in the tree yet.
		// This is expected, and there is nothing more to check so return a zero Attestations state.
		return &Attestations{}, nil
```

**Review by wlynch — APPROVED** [review]
One small comment tweak, but otherwise lgtm (feel free to submit)

File: internal/attestations/attestations.go (hunk 9 above)
Comment (adityasaky) [reply internal/attestations/attestations.go:17]: I see #237 is open to track this!
